app/src/main/java/org/mozilla/fenix/HomeActivity.kt

```diff
@@ -236,13 +244,28 @@ open class HomeActivity : LocaleAwareAppCompatActivity(), NavHostActivity {
     }
 
     protected open fun startupTelemetryOnCreateCalled(safeIntent: SafeIntent, hasSavedInstanceState: Boolean) {
-        components.appStartupTelemetry.onHomeActivityOnCreate(safeIntent, hasSavedInstanceState)
+        components.appStartupTelemetry.onHomeActivityOnCreate(
+            safeIntent,
+            hasSavedInstanceState, homeActivityInitTimeStampNanoSeconds
+        )
+
+        OneShotPreDrawListener.add(rootContainer) {
```

Comment:
Wouldn't it be better to call this in `onHomeActivityOnCreate`? That way we're not splitting up the logic and minimizing low-level code in an Activity class.

Also, `rootContainer.doOnPreDraw { ... }` is more natural than this invocation style.

Reply:
that would require us passing in the rootContainer inside the `onHomeActivityOnCreate`, which I was trying to avoid since I wanted to avoid passing in UI elements to the data class. maybe I misunderstood the suggestion?

Also `OneShotPreDrawListener` removes the listener after the first call, which is what we need. but if we were to use `rootContainer.doOnPreDraw { ... } `then we have the overhead of removing the listener ourself.

Reply:
> which I was trying to avoid since I wanted to avoid passing in UI elements to the data class.

By data class, do you mean you wanted to avoid passing UI elements into the model? If so, that's an interesting point – I haven't been conceptualizing consider the split between model and UI in the app startup telemetry class. I've been thinking of it as an extension of the Activity – its responsibility is to record startup data, primarily by managing arguments and lifecycles to do it, i.e. no separation between model and UI. In my model, it makes sense to put ALL of the logic in App Startup Telemetry (or files only app startup telemetry references), to encapsulate it.

Plus, as you know, I hate adding more than a line or so per PR to Activity and Fragment methods. :)

But I'll let you decide – I'm not sure how you're designing the class.

> but if we were to use rootContainer.doOnPreDraw { ... } then we have the overhead of removing the listener ourself.

`doOnPreDraw` uses OneShot... under the hood so I don't think that's true. Note: a design philosophy of KTX functions (this is one) is that they never add functionality over what's in the framework/androidx: they just make it more natural to access in Kotlin.

Reply:
Ah my bad! I mistook the `doOnPreDraw` listener for `ViewTreeObserver.OnPreDrawListener `.

I changed `oneShotListener` to `doOnPreDraw`. 

As for thinking of `AppStartupTelemetry` as an extension, thats a pretty good way of thinking about it and I agree with it. 
However, in this case, since we are adding a listener on, I wonder if the pros (4 less line in `homeactivty`) outweighs the cons (unit testing a listener inside a function). I haven't looked at how complicated it is but I'm also more inclined to keep testing simple?

Reply:
> However, in this case, since we are adding a listener on, I wonder if the pros (4 less line in homeactivty) outweighs the cons (unit testing a listener inside a function). I haven't looked at how complicated it is but I'm also more inclined to keep testing simple?

What do you mean by unit testing a listener inside a function?

fwiw, I think we should consider the incidental complexity that occurs when we're leaving the code here:
- we're making HomeActivity longer and more complex.
- we're making AppStartupTelemetry more complex by splitting its functionality into multiple files

And at a more nuanced, low-level:
- in the future, it's more likely that if this method is modified, HomeActivity will be coupled to this listener. 
- we're duplicating the `onPreDraw` call here (e.g. I noticed you didn't use `doOnPreDraw` in `ExternalAppBrowserActivity`). if we fix it with a helper function, we're making HomeActivity more complicated even though we already have a deduplication mechanism inside the AppStartupTelemetry (so there would be duplication of the deduplication mechanism)